homeassistant/helpers/entity.py

```diff
@@ -595,6 +670,15 @@ def _async_write_ha_state(self) -> None:
         if (icon := (entry and entry.icon) or self.icon) is not None:
             attr[ATTR_ICON] = icon
 
+        if (audio_urls := self.audio_urls) is not None:
+            attr[ATTR_AUDIO_URLS] = audio_urls
+
+        if (image_urls := self.image_urls) is not None:
+            attr[ATTR_IMAGE_URLS] = image_urls
+
+        if (video_urls := self.video_urls) is not None:
+            attr[ATTR_VIDEO_URLS] = video_urls
```

Comment:
This is adding a lot of data to the state machine that is rarely used. And this data is also not part of the state, it's part of what they saw in the past. For motion detection events, this can become thousands of events. The reason for media source we went as an API is because we can't just dump it all in the state machine.

Reply:
A motion event is not a singular point in time. Motion detection has state, at least with UniFi Protect cameras. Usually the duration is rather short (~10-30 seconds), but it still has a period of time. There is a point when motion is detected and there is a point when motion is no longer detected. Same with objection detection, there is a point when a camera detects a person and a point the camera stops detecting that person. 

The data is also excluded from recorder so it should only remain in memory, unless I am missing something. Making the assumption it will be "rarely used" is projecting your use case on everyone else. I have seen users that use motion detections to detect when their cats eat their food, use motion events to detect when their child wakes up, or use events for monitoring remote locations where any amount of motion is cause for alert. When you add in object detection like via Frigate, Blue Iris or UniFi Protect, then all detection events are notable. I have notifications for all of my outdoor cameras whenever they detect a person or a car with the public sideway and street excluded. 

Additionally, the API is not available to templating/automation engines for users to fetch arbitrary media items. Using `entity_picture` on the camera entity is also not adequate since even a matter of a few frames can means getting a bad image. Also, integration (like UniFi Protect) can generate cropped images focused on the target of the detection event instead of full sized 4096x2160 image that really makes for a poor image for a smart phone notification.

Exposing thumbnail URLs at time of event so the Home Assistant automation/template engine and NodeRed can access them for sending notifications is literally the most requested feature for the UniFi Protect integration.

Reply:
This shouldn't exist in the base entity class. It should get its own base class instead. Most entities aren't going to use these properties.

Reply:
Making it separate base class makes sense, but since everything already inherits from `Entity`, that would mean integration devs would need to do multiple inheritance (`MediaEntity`, `SwitchEntity` as an example). Do you think that would lead to issues?

Reply:
We do multiple inheritance in a few places.  It's sometimes a bit tricky but in general it works just fine.  Switchbot has a few that are 5 deep

Reply:
I think multiple inheritance would solve my cyclic import issue as well. I can put the `MediaEntity` in the `media_source` integration, we should greatly simplify that set up. Will refactor to that shortly.

Reply:
The state machine is not a data store. We've been moving more data out of it to things like entity/device/area registry. We shouldn't add things.

We have a couple of use cases where we want to get async information about something. For example prediction of what data will be like in X time or the next 2 hours of calendar events.

Ideas have been floating around to allow defining data sources that would be evaluated before running a template. A bit like a `variables` block in a script, but allow them to be async.

Could look like this:

```yaml
template:
  - trigger:
      - platform: state
        entity: binary_sensor.kitchen_motion
        to: "on"
    data_source:
      total_motion:
        platform: binary_sensor
        source: total_on
        entity_id: binary_sensor.kitchen_motion
        duration: 3600
    sensor:
      - name: "Motion in last hour"
        state: '{{ total_motion / 60 | round }}'
        unit_of_measurement: "Minutes"
```

This kind of thing would actually solve many different async use cases.